### PR TITLE
feat: allow fuzzy invalidation

### LINKED ISSUE #3648: feat: query invalidation should take `Partial<Type>` instead of `Type` as the input
### Describe the feature you'd like to request

`react-query` matches query invalidation in a fuzzy way, and allows to invalidate queries without having the whole input used to call it.

From this comment on the react query forums: https://github.com/TanStack/query/discussions/4762#discussioncomment-4649876
> from a pure react-query perspective, if your `queryKey` is:
> 
> ```
> ['my-path', { a: 'b', foo: 'bar' }]
> ['my-path', { a: 'b', something: 'else' }]
> ```
> 
> it should be possible to target them both for query invalidation by doing:
> 
> ```
> queryClient.invalidateQueries({
>   queryKey: ['my-path', { a: 'b' }]
> })
> ```
> 
> because that does fuzzily match them both. It might be that trpc doesn't have a concept for that though ...

Trpc does not currently allow to match queries this way.

### Describe the solution you'd like to see

Having a query with the input type type:

``` typescript
type TInput = { id: string, name: string, lastName: string | undefined } 
```
Even though we call the query with the full TInput, the invalidation should allow to do this:
``` typescript
query.invalidate({id: 'id'})
```
This should be easy to do, as `react-query` does already behave like that, it's just a type issue. It should be a deep partialization of it.

### Describe alternate solutions

Making it `Partial<TInput>` would solve some problems, but I believe `react-query` will match everything fuzzily so the deep partialization is better.

### Additional information

_No response_

### 👨‍👧‍👦 Contributing

- [X] 🙋‍♂️ Yes, I'd be down to file a PR implementing this feature!

### PR DESCRIPTION
**Pull request by nicolassanmar**
Closes #3648

## 🎯 Changes

Allows using the `.invalidate` function with partial types for the input. This allows to fuzzily match the queries when not using `exact: true` as an option.

## ✅ Checklist

- [x] I have followed the steps listed in the [Contributing guide](https://github.com/trpc/trpc/blob/main/CONTRIBUTING.md).
- [ ] If necessary, I have added documentation related to the changes made.
- [x] I have added or updated the tests related to the changes made.

### CODE REVIEW

File: packages/react-query/src/shared/proxy/utilsProxy.ts [hunk 1]
```diff
@@ -32,6 +32,12 @@ import {
 import { TRPCContextState } from '../../internals/context';
 import { getQueryKey } from '../../internals/getQueryKey';
 
+type DeepPartial<TObject> = TObject extends object
+  ? {
+      [P in keyof TObject]?: DeepPartial<TObject[P]>;
+    }
+  : TObject;
+
```
Comment (nicolassanmar) [inline packages/react-query/src/shared/proxy/utilsProxy.ts:40]: Where would you put this utility type?

Comment (juliusmarminge) [reply packages/react-query/src/shared/proxy/utilsProxy.ts:40]: I think we have most in trpc/server in some types.ts file

Comment (nicolassanmar) [reply packages/react-query/src/shared/proxy/utilsProxy.ts]: Thanks! I moved the util type there

**Comment by bautistaaa** [comment]
should add a test for this

**Comment by juliusmarminge** [comment]
> should add a test for this

Yea add a test where the filter isnt exact and make sure its ok for TS + that it invalidates properly 👍👍

**Review by KATT — CHANGES_REQUESTED** [review]
Please add a test so we don't accidentally break this in the future :)

**Review by juliusmarminge — APPROVED** [review]

**Review by KATT — APPROVED** [review]

**Review by KATT — APPROVED** [review]

**Pull request merged**